src/backend/distributed/connection/connection_configuration.c

```diff
@@ -0,0 +1,239 @@
+	ConnParams.values[ConnParams.size] = strdup(value);
+	ConnParams.size++;
+
+	ConnParams.keywords[ConnParams.size] = ConnParams.values[ConnParams.size] = NULL;
+}
+
+
+bool
+CheckConninfo(const char *conninfo, const char **whitelist,
+			  Size whitelistLength, char **errmsg)
+{
+	PQconninfoOption *optionArray = NULL;
+	PQconninfoOption *option = NULL;
+	Index whitelistIdx PG_USED_FOR_ASSERTS_ONLY = 0;
+
+	if (errmsg == NULL)
+	{
+		char *errmsgString;
+		errmsg = &errmsgString;
+	}
+
+	if (conninfo == NULL)
+	{
+		return true;
+	}
+
+	if (uri_prefix_length(conninfo) != 0)
+	{
+		return false;
+	}
+
+	optionArray = PQconninfoParse(conninfo, NULL);
+
+	if (optionArray == NULL)
+	{
+		return false;
+	}
+
+	optionArray = PQconninfoParse(conninfo, NULL);
+
+	if (optionArray == NULL)
+	{
+		return false;
+	}
+
+#ifdef USE_ASSERT_CHECKING
+	for (whitelistIdx = 1; whitelistIdx < whitelistLength; whitelistIdx++)
+	{
+		const char *prev = whitelist[whitelistIdx - 1];
+		const char *curr = whitelist[whitelistIdx];
+
+		AssertArg(strcmp(prev, curr) < 0);
+	}
+#endif
+
+	for (option = optionArray; option->keyword != NULL; option++)
+	{
+		bool invalidKeyword = true;
+
+		if (option->val == NULL || option->val[0] == '\0')
+		{
+			continue;
+		}
+
+		invalidKeyword = bsearch(&option->keyword, whitelist, whitelistLength,
+								 sizeof(char *), pg_qsort_strcmp) == NULL;
+
+		if (invalidKeyword)
+		{
+			*errmsg = "Invalid conninfo keyword detected.";
+			PQconninfoFree(optionArray);
+			return false;
+		}
+	}
+
+	PQconninfoFree(optionArray);
+
+	return true;
+}
+
+
+void
+GetConnParams(ConnectionHashKey *key, char ***keywords, char ***values,
+			  MemoryContext context)
+{
+	/* sign, 10 digits, NUL */
+	char *nodePortString = MemoryContextAlloc(context, 12 * sizeof(char *));
+	const char *runtimeKeywords[] = {
+		"host", "port", "dbname", "user", "client_encoding"
+	};
+	const char *runtimeValues[] = {
+		MemoryContextStrdup(context, key->hostname),
+		nodePortString,
+		MemoryContextStrdup(context, key->database),
+		MemoryContextStrdup(context, key->user),
+		GetDatabaseEncodingName()
+	};
+	Index authParamsIdx = ConnParams.size + lengthof(runtimeKeywords);
+
+	pg_ltoa(key->port, nodePortString);
+
+	*keywords = MemoryContextAllocZero(context, ConnParams.maxSize * sizeof(char *));
+	*values = MemoryContextAllocZero(context, ConnParams.maxSize * sizeof(char *));
+
+	memcpy(*keywords, ConnParams.keywords, ConnParams.size * sizeof(char *));
+	memcpy(*values, ConnParams.values, ConnParams.size * sizeof(char *));
+
+	memcpy(&((*keywords)[ConnParams.size]), runtimeKeywords, sizeof(runtimeKeywords));
+	memcpy(&(*values)[ConnParams.size], runtimeValues, sizeof(runtimeValues));
+
+	(*keywords)[authParamsIdx] = (*values)[authParamsIdx] = NULL;
+}
+
+
+static Size
+CalculateMaxSize()
+{
+	PQconninfoOption *defaults = PQconndefaults();
+	PQconninfoOption *option = NULL;
+	Size maxSize = 0;
+
+	for (option = defaults, maxSize = 1;
```

Comment:
I get that we need to allocate the number of possible keywords + 1 because the last entry should be NULL. However, doing the +1 by initialising maxSize to 1 here is way too subtle. Can we just count the number of keywords here and add 1 at the call site?

Reply:
Well I mean, you figured it out 😆!

Sure, I'll do that.